### PR TITLE
kubeadm: add guide for kubelet serving certs and metrics-server

### PR DESCRIPTION
**Pull request by neolit123**
**note: targeting the `master` branch is accurate for this change.**

Include a new section of the kubeadm certificate management page
to talk about kubelet serving certificates and how to make
them signed. Also include a note about using secure connection
with the metrics-server.

fixes https://github.com/kubernetes/kubeadm/issues/1602

/sig cluster-lifecycle

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md [hunk 1]
```diff
@@ -236,3 +236,65 @@ After a certificate is signed using your preferred method, the certificate and t
...
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+---
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+serverTLSBootstrap: true
+```
+
+If you have already created the cluster you must adapt it by doing the following:
+- Find and edit the `kubelet-config-x.yy` ConfigMap under `kube-systedm` to include
```
Comment (jdef) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:260]:
```suggestion
- Find and edit the `kubelet-config-x.yy` ConfigMap under `kube-system` to include
```
typo?

File: content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md [hunk 2]
```diff
@@ -415,3 +415,14 @@ If `/var/lib/kubelet` is being mounted, performing a `kubeadm reset` will effect
 To workaround the issue, re-mount the `/var/lib/kubelet` directory after performing the `kubeadm reset` operation.
 
 This is a regression introduced in kubeadm 1.15. The issue is fixed in 1.20.
+
+## Cannot use the metrics-server securely in a kubeadm cluster
+
+In a kubeadm cluster, the [metrics-server](https://github.com/kubernetes-sigs/metrics-server)
+can be used insecurely by passing the `--kubelet-insecure-tls` to it. In most cases this is not recommended.
```
Comment (sftim) [inline content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md:422]: Let's reword this to always recommend TLS for traffic between this add-on and kubelets. Is there ever a case when we don't recommend it?

File: content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md [hunk 3]
```diff
@@ -415,3 +415,14 @@ If `/var/lib/kubelet` is being mounted, performing a `kubeadm reset` will effect
 To workaround the issue, re-mount the `/var/lib/kubelet` directory after performing the `kubeadm reset` operation.
 
 This is a regression introduced in kubeadm 1.15. The issue is fixed in 1.20.
+
+## Cannot use the metrics-server securely in a kubeadm cluster
```
Comment (sftim) [inline content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md:419]: I'm not sure about this title. Clearly you _can_ use the metrics-server in a kubeadm cluster, including encryption in transit.

What's the warning that someone sees if they haven't read this advice? Could we use that warning as the basis for the title?

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md [hunk 4]
```diff
@@ -236,3 +236,65 @@ After a certificate is signed using your preferred method, the certificate and t
 Kubeadm does not support rotation or replacement of CA certificates out of the box.
 
 For more information about manual rotation or replacement of CA, see [manual rotation of CA certificates](/docs/tasks/tls/manual-rotation-of-ca-certificates/).
+
+## Enabling signed kubelet serving certificates {#kubelet-serving-certs}
+
+By default the kubelet serving certificate deployed by kubeadm is self-signed.
+This means a connection from external services like the
+[metrics-server](https://github.com/kubernetes-sigs/metrics-server) to a
+kubelet cannot be not be secured.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:245]: Could we mention “TLS” here?

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md [hunk 5]
```diff
@@ -236,3 +236,65 @@ After a certificate is signed using your preferred method, the certificate and t
...
+If you have already created the cluster you must adapt it by doing the following:
+- Find and edit the `kubelet-config-x.yy` ConfigMap under `kube-systedm` to include
+the field `serverTLSBootstrap: true` under the `config` key
+- On each node, add the `serverTLSBootstrap: true` field in `/var/lib/kubelet/config.yaml`
+and restart the kubelet with `systemctl restart kubelet`
+
+The field `serverTLSBootstrap: true` will enable the bootstrap of kubelet serving
+certificates by requesting them from the `certificates.k8s.io` API. One known limitation
+is that the CSRs (Certificate Signing Requests) for these certificates cannot be automatically
+approved by the default signer in kube-controller-manager. This will require action from
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:268]: Which default signer? If possible, link to https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers and specifiy which built in signer we mean, ie `kubernetes.io/kubelet-serving`.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md [hunk 6]
```diff
@@ -236,3 +236,65 @@ After a certificate is signed using your preferred method, the certificate and t
...
+to expiration a new set of CSRs for the serving certificates will be created and must
+be approved to complete the rotation. To understand more see
+[Certificate Rotation](/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/#certificate-rotation).
+
+If you are looking for a solution for automatic approval of these CSRs it is recommended
+that you contact your cloud provider and ask if they have a CSR signer that verifies
+the node identity with an out of band mechanism.
+
+Third party custom controllers such as
+[kubelet-rubber-stamp](https://github.com/kontena/kubelet-rubber-stamp) can also be used.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:296]: Even though it's one item, I'd make this be a list of links and include the third party content warning:
`{{% thirdparty-content %}}`

before that list.

**Review by sftim — COMMENTED** [review]
Thanks!

Some feedback. I'm not sure about the heading “Cannot use the metrics-server securely in a kubeadm cluster” and also I'd like to include a 3rd party content warning for `kubelet-rubber-stamp`.

Plus a few less important details.

File: content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md (hunk 2 above)
Comment (neolit123) [reply content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md:422]: the metrics server docs state the insecure can be used for testing:
> --kubelet-insecure-tls - Do not verify the CA of serving certificates presented by Kubelets. For testing purposes only.

https://github.com/kubernetes-sigs/metrics-server/blob/master/README.md#configuration

i can reword to:
> This is not recommended for production clusters.

File: content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md (hunk 3 above)
Comment (neolit123) [reply content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md:419]: as per the kubeadm default setup, one can only use the metrics-server if they pass `--kubelet-insecure-tls`.

due to the default self-signed serving certificate for the kubelet they will get errors like:

> x509: certificate signed by unknown authority
> x509: certificate is valid for IP-foo not IP-bar

we can include these two as symptoms.

even if the certificate is properly signed, i don't know what errors happen later if TLS related flags are not properly configured for the metrics server.

the title seems fine to me, so if you think something is better could you recommend an exact replacement?

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md (hunk 4 above)
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:245]: sure, there is also a typo there it seems.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md (hunk 5 above)
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:268]: i can mention `kubernetes.io/kubelet-serving`

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md (hunk 6 above)
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:296]: do you mean something in the lines of?

```
{{% thirdparty-content %}}
Third party custom controllers can also be used:
- [kubelet-rubber-stamp](https://github.com/kontena/kubelet-rubber-stamp)
{{% /thirdparty-content %}}

Such a controller is not a secure mechanism unless it not only verifies the CommonName
...
```

**Comment by neolit123** [comment]
updated to address the review comments.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md [hunk 7]
```diff
@@ -236,3 +236,69 @@ After a certificate is signed using your preferred method, the certificate and t
...
+
+{{% thirdparty-content %}}
+
+Third party custom controllers can be used:
+- [kubelet-rubber-stamp](https://github.com/kontena/kubelet-rubber-stamp)
+
+Such a controller is not a secure mechanism unless it not only verifies the CommonName
+in the CSR but also verifies the requested IPs and domain names. This would prevent
+a malicious actor that has access to a kubelet client certificate to create
+CSRs requesting serving certificates for any IP or domain name.
```
Comment (fabriziopandini) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:304]: I don't think we should express options on external project; TBH I would avoid the link entirely but I don't want to block on this

Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:304]: @sftim and randomvariable seem to be fine with mentioning this third-party option.

for cloud provider scenarios we already tell users to consult their provider.
users that don't trust this would have to either use rubber-stamp or write their own controller / DS / scripts.

**Comment by alvaroaleman** [comment]
@neolit123: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by sftim** [comment]
Needs a rebase, but assuming the obvious approach to rebasing: LGTM.

**Comment by neolit123** [comment]
rebased.

/cc @fabriziopandini 
for tech review / LGTM.

File: content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md [hunk 8]
```diff
@@ -420,3 +420,20 @@ To workaround the issue, re-mount the `/var/lib/kubelet` directory after perform
...
+
+If you want to use TLS between the metrics-server and the kubelet there is a problem,
+since kubeadm deploys a self-signed serving certificate for the kubelet. This can cause the following errors
+on the side of the metrics-server:
+```
+x509: certificate signed by unknown authority
+x509: certificate is valid for IP-foo not IP-bar
+```
+
+See [Enabling signed kubelet serving certificates](/docs/tasks/administer-cluster/kubeadm/kubeadm-certs#kubelet-serving-certs)
```
Comment (sftim) [inline content/en/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm.md:436]: nit (OK to merge without addressing):
```suggestion
See [Enabling signed kubelet serving certificates](/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs)
``

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md [hunk 9]
```diff
@@ -207,3 +207,69 @@ After a certificate is signed using your preferred method, the certificate and t
...
+kind: ClusterConfiguration
+---
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+serverTLSBootstrap: true
+```
+
+If you have already created the cluster you must adapt it by doing the following:
+- Find and edit the `kubelet-config-x.yy` ConfigMap under `kube-system` to include
+the field `serverTLSBootstrap: true` under the `config` key
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:232]: We could also write:
```suggestion
 - Find and edit the `kubelet-config-x.yy` ConfigMap in the `kube-system` namespace.
   In that ConfigMap, the `config` key has a
   [kubelet configuration](/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration) document as
   its value. Edit the kubelet configuration document to set `serverTLSBootstrap: true`.
```

- maybe in a follow-up PR?

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md:232]: https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/ recommends changing the ConfigMap that kubelets use (a different name); is that advice related to what's recommended here?

**Review by sftim — COMMENTED** [review]
Some feedback. I don't think it need block merging these changes.

LGTM!

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md (hunk 9 above)
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md]: > maybe in a follow-up PR?

updated in this PR.

> https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/ recommends changing the ConfigMap that kubelets use (a different name); is that advice related to what's recommended here?

kubeadm uses a single ConfigMap, so need to introduce a new ConfigMap with a different name.

**Comment by neolit123** [comment]
@sftim 
updated with your two suggestions above.
tech LGTM already passed by @fabriziopandini

**Comment by kbhawkey** [comment]
Page previews:
https://deploy-preview-27071--kubernetes-io-master-staging.netlify.app/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs
https://deploy-preview-27071--kubernetes-io-master-staging.netlify.app/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm/#cannot-use-the-metrics-server-securely-in-a-kubeadm-cluster
